### DISCUSSION TITLE
identifying changes and merging mindmaps

### DISCUSSION THREAD

**Discussion by nnako** [freeplane/freeplane#496] [Help]
**BACKGROUND**

Mindmaps are great means to share structured information. Challenges arise if there is more than one person working on a particular file. Especially when the mindmap already has grown to a certain size and keeping track of changes becomes difficult.

**USE CASE**

I want to work in a big mindmap (add, delete, move, modify nodes as well as certain styles) together with a handful of other people. In contrast to the "[easy collaboration idea](https://github.com/freeplane/freeplane/discussions/426#discussion-4120709)", where people might work `simultaneously` within the same mindmap, I want to realize a `sequential` editing concept. Working sequentially within a mindmap requires some kind of "branching" and "merging" as done regularly when working on code with multiple developers.

Now, when trying to merge changes made by another person back into a "main" mindmap branch, classical `diffing` as provided by all the `git` tools out there reach their usability limit. Those tools operate line-based, not structure-based. This leads to thousands of "changes" regarding all kinds of information:

- actual core text changes
- color or style metadata
- modification date changes
- folding state of nodes
- various structure modifications without user text changes (e.g. moved arrow links)
- ...

here is an actual diff example of the `same` mindmap file where changes have gone wild (on the left you see the "mini map" and on the right you see the actual changes which have to be evaluated and copied between the two content sides):

![image](https://user-images.githubusercontent.com/5845602/178770973-c9173977-b60e-4ab1-873e-76f46bddd71a.png)

only the changes of "folding" are shown as differences. There seems to be no way to only show actual textual changes, rendering the folding changes invisible to the merging person. At least not with these "standard" text comparison tool.

**QUESTION**

How do you deal with this use case of `merging` mindmap files? How do you only see "relevant" e.g. core text changes?

There might be even ways to compare the two mindmaps within the Freeplane editor itself, but even then you would have to have some kind of agreement between the people working within this mindmap. Like "if you make a change, set a certain `style` attribute". But as the mindmaps grow, filtering those `styles` in order to identify changes becomes stressful.

Any ideas? I would be interested in knowing how you deal with merging Freeplane mindmaps after changes.

**Comment by EdoFro** [comment]
As a first help, install this addon:

https://github.com/EdoFro/Freeplane_Map-Saving-Options

> # Description
> 
> **'Map Saving Options'** is a **Freeplane AddOn** that gives you the possibility to add **saving preferences** to **each** of your **maps**.
> 
> Mindmaps can be used for different purposes and saving information like node's modification times depends most of the times on the map itself, its content and use case.
> 
> With this AddOn you can define if a map have to save the information of any of these:
> 
> * folded and unfolded nodes
> * selected node when saving
> * *Creation* and *Modification times* of each node
> 
> This way, you can define for each map if it will use the **central Freeplane preferences**, its own or a mix of both.

**Reply by EdoFro** [reply]
> * selected node when saving

this is not saved in the map, it is saved in the user preferences: So it doesn't affect when comparing two mindmap files.

**Comment by EdoFro** [comment]
I was thinking also in create an addon that helps to "merge" two mindmaps.

My idea is to simulate the git process in freeplane itself. I have some test scripts that compare two versions of a node to discover their differences.
I think using this logic the user can select which changed nodes he/she wants to commit and which ones not. This way you can have a common updated mindmap, but each one can have additional personal information in its own version. 

It's still an idea, but if somebody else is interested I can create a github repository where we can develop this.

**Reply by nnako** [reply]
Hi @EdoFro , thanks for all your concept suggestions. I would think that a solution would reach the maximum number of users if it was built-in into Freeplane and thus harmonizable with other features.

* A kind of visual left-side-right-side approach as seen e.g. in WinMerge or other git-like approaches
* where the addon provides appropriate filter options to identify only the "relevant" changes
* and where the user might then step through all these changes (on either side) and decide about "keep" or "discard"
* where the addon keeps an eye on the correctness of the underlying XML structure to never corrupt it
* ...
* A 2nd development milestone might be the integration of git version management (see [Collaborative-Addon](https://github.com/seb4stien/freeplane-collaborative-addon) by seb4stien, about ten years ago)
* A 3rd development milestone could be to combine it with the [simple collaboration feature](https://github.com/freeplane/freeplane/discussions/426) I introduced some months ago

Yes, please create the github repo and we can continue there, if you like.

**Reply by EdoFro** [reply]
Ok. I will create the repo.

Here is a first version of what I'm currently using when I have two versions of a map and want to see their differences and combine them. (I prefer this way than the XML comparision)

It is a script that has a list of tests that I comment or uncomment depending on what node's properties I think are important when comparing the two maps.
Both maps should be open in Freeplane and their names should start with the same string (in the script I used 'map', but you have to change that for your own maps).

This script travels the maps recursively.
It is an iterative process: the scripts stops when it discover the first difference, then I manually decide which is the correct version and make them equal on both maps. Then I run the script again to discover the next difference.

Another way to do a comparision is to compare nodes with the same ID on both maps: comparing their properties, parent's id, and children id's to discover nodes and branches that have been moved (and where they have been moved to). (NOT in this script)

```groovy
//list of tests to be performed by the comparator routine.
// You can add more or comment the ones you don't need (in my case it depends on the maps)
tests = [
          [ ' .text                        ' ,{a,b -> a.text == b.text}                                           ]
      //   ,[ ' .id                          ' ,{a,b -> a.id == b.id}                                               ]
         ,[ ' .details?.htmlText           ' ,{a,b -> a.details?.htmlText == b.details?.htmlText}                 ]
         ,[ ' .note?.html                  ' ,{a,b -> a.note?.html == b.note?.html}                               ]
         ,[ ' .children.size()             ' ,{a,b -> a.children.size() == b.children.size()}                     ]
         ,[ ' .children*.text              ' ,{a,b -> a.children*.text == b.children*.text}                       ]
         ,[ ' .attributes.size()           ' ,{a,b -> a.attributes.size() == b.attributes.size()}                 ]
         ,[ ' .attributes.names.sort()     ' ,{a,b -> a.attributes.names.sort() == b.attributes.names.sort()}     ]
         ,[ ' .attributes.getAll(nom)      ' ,{a,b -> (a.attributes.names + b.attributes.names).unique().sort().every 
                                                           {  a.attributes.getAll(it).sort() == b.attributes.getAll(it).sort() }
                                                         }]
         ,[ ' .style.name                  ' ,{a,b -> a.style.name == b.style.name}                               ]
         ,[ ' .icons.icons.sort()          ' ,{a,b -> ([] + a.icons.icons).sort() ==([] + b.icons.icons).sort()}  ] //icons (no specific order)
         ,[ ' .icons.icons                 ' ,{a,b -> a.icons.icons == b.icons.icons}                             ] //icons in specific order
         ,[ ' .parent.getChildPosition(a)  ' ,{a,b -> a.isRoot() || (a.parent.getChildPosition(a) == b.parent.getChildPosition(b)) }  ]
]

isDifferentAttr = 'isDifferent'

def (a,b) = rootsFromOpenMapsWhichNamesStartWith( 'map' ) //change 'map' for a string that suites your maps to be compared

a.find{n -> n.attributes.containsKey(isDifferentAttr)}.each{n-> n[isDifferentAttr] = null}
b.find{n -> n.attributes.containsKey(isDifferentAttr)}.each{n-> n[isDifferentAttr] = null}

def mapsAreEqual = comparar(a,b)

ui.informationMessage( "${mapsAreEqual? 'No different nodes' : 'Difference in node'} encountered in compared maps" )



// ----methods---------------------------------------------------------------

def comparar(x,y){
    print "${x.text} --> "
    if(equals(x,y)){
        def xChilds = x.children
        def yChilds = y.children
        for(def i=0 ; i< xChilds.size(); i++){
            def resp = comparar(xChilds[i],yChilds[i])
            if(!resp){
                return false
            }
        }
        return true
    }else{
        def resp = equals(x,y, true)
        x.pathToRoot.dropRight(1).reverse()*.folded = false
        y.pathToRoot.dropRight(1).reverse()*.folded = false
        x[isDifferentAttr] = resp
        y[isDifferentAttr] = resp
        x.attributes.optimizeWidths()
        y.attributes.optimizeWidths()
        //c.select([x,y])
        return false
    }
}

def equals(x,y,doLogPrint = false){
    def test = true
    def i = 0
    def iMax = tests.size()
    while(test && i<iMax){
        test = tests[i][1](x,y)
        if(doLogPrint){
            println i + '.- ' + tests[i][0].toString() + '   ' + test
        }
        i++
    }
    println i==iMax
    return !doLogPrint ? test : tests[i-1][0].toString()
}

def rootsFromOpenMapsWhichNamesStartWith(nameStart){
    def (map1,map2) = c.openMaps.findAll{it.name.startsWith(nameStart)}.take(2)
    //present to user which are the selected maps
    def texto1 = """Maps to be compared
  map 1 :
    '${map1?.name}'
    ${map1?.file?.path}
    
  map 2 :
    '${map2?.name}'
    ${map2?.file?.path}
    
    """
    
    uimsgx(texto1)

    return [map1.root, map2.root]
}

def uimsgx(msg){
    assert 2 != ui.showConfirmDialog(null, msg.toString(),'Continue?',2)
}

```

And here are two example maps to play with:
[exampleMaps.zip](https://github.com/freeplane/freeplane/files/9446277/exampleMaps.zip)

bye,

edo

**Reply by EdoFro** [reply]
I changed the `comparar `method so that it shows the nodes that have the difference:

```groovy
def comparar(x,y){
    print "${x.text} --> "
    if(equals(x,y)){
        def xChilds = x.children
        def yChilds = y.children
        for(def i=0 ; i< xChilds.size(); i++){
            def resp = comparar(xChilds[i],yChilds[i])
            if(!resp){
                return false
            }
        }
        return true
    }else{
        def resp = equals(x,y, true)
        
        c.mapLoader(x.mindMap.file).withView().getMindMap()
        sleep(500)
        x.pathToRoot.dropRight(1).reverse()*.folded = false
        x[isDifferentAttr] = resp
        x.attributes.optimizeWidths()
        c.centerOnNode(x)
        c.select(x)
        sleep(500)
        
        c.mapLoader(y.mindMap.file).withView().getMindMap()
        sleep(500)
        y.pathToRoot.dropRight(1).reverse()*.folded = false
        y[isDifferentAttr] = resp
        y.attributes.optimizeWidths()
        c.centerOnNode(y)
        c.select(y)
        sleep(500)
        
        return false
    }
}
```

if you have a view with both maps side by side then it looks like this:
before comparing:
![image](https://user-images.githubusercontent.com/35700575/187266631-46814fb3-8352-4a5d-a652-75f5e29acb6f.png)

after running the script:
![image](https://user-images.githubusercontent.com/35700575/187267000-8e772f8d-5e3a-448e-8120-11c63041a981.png)

In this case it recognized that the nodes had the same icons but in a different order

**Comment by EdoFro** [comment]
Another idea is using a script that compairs both XML .mm files

more info here:
https://groovy-lang.org/processing-xml.html

**Comment by EdoFro** [comment]
And my last idea is using a software to compare and merge files. I use [WinMerge](https://winmerge.org/) and it has the option to use filters, plugins, unpackers, prediff, scripts, etc. to customize how two files should be compared. I don't know how to create a custom process for .mm files. I use it a lot but just the standard features.

Hope it helps

**Reply by EdoFro** [reply]
I made a test with [Winmerge](https://winmerge.org/) and it worked pretty good by doing these steps:

1. load both mindmap versions (Menu File / Open)

![image](https://user-images.githubusercontent.com/35700575/178841993-ca92ef42-ccd8-4509-a20e-ccd00f841c88.png)

2. and click "Compare" and it shows:

![image](https://user-images.githubusercontent.com/35700575/178842120-cdf643d4-a376-4f36-83b5-a80e385cd18c.png)

3. apply PrettifyXML:

![image](https://user-images.githubusercontent.com/35700575/178843338-7fcff076-ede7-433b-ae09-05f8bfc4f325.png)

4. Now it looks like this:

![image](https://user-images.githubusercontent.com/35700575/178842496-d08996bc-e9f5-4208-9561-f3034c03d141.png)

5. open the Plugins dialog (menu Plugins / Plugins settings...)
    - select PrediffLineFilter
    - change File filters to "\.mm$"
    - and click button "Plugin Settings..."
 
![image](https://user-images.githubusercontent.com/35700575/178843658-2129a1fe-e6b7-4351-b2e0-bfea761f0d6b.png)

6. add some regex expresions to define lines to be ignored
    - for example `\bFolded="*."\B` to ignore all changes related to folding nodes

Here is an image of the parameters of my test:
![image](https://user-images.githubusercontent.com/35700575/178844236-f7bf3f06-9ad6-474c-b42a-d77344e2e7f4.png)

7. Apply the PrediffLineFilter plugin:

![image](https://user-images.githubusercontent.com/35700575/178844871-2551292e-4215-4d64-939a-ea393af5e2d7.png)

8. Ready! Now it shows the ignored changes in a pale yellow color and it counts and navigates only the not ignored changes.

![image](https://user-images.githubusercontent.com/35700575/178845322-302c453b-398c-47ed-b38f-54ac322b8507.png)

Hope it helps!
edo

**Reply by nnako** [reply]
Hi @EdoFro . Nice job! Thanks. I think, you nailed it by showing your workflow. I will check it for myself, soon.

If there was a way to "automate" the sequence of operation you showed in your post, the solution would be ready. Currently, I don't know if WinMerge provides a command line interface by which choosing and configuring it's plugins could be done. If it does, an automated solution for merging mindmaps by regarding only desired portions of content would be easily possible.

Let us see.

**Reply by nnako** [reply]
just found some documentation about automating WinMerge for this:

- [Using WinMerge with other tools](https://manual.winmerge.org/en/Version_control.html#:~:text=The%20WinMerge%20command%2Dline%20interface,to%20launch%20a%20WinMerge%20operation.)
- [Command line interface specification](https://manual.winmerge.org/en/Command_line.html)

**Reply by nnako** [reply]
Hi again, @EdoFro . Unfortunately, there seem none of your WinMerge plugins available when I open my plugins menu:

![image](https://user-images.githubusercontent.com/5845602/186976011-64fa5dbb-d5aa-4764-81a9-75c0d22e8a3a.png)

Have you installed those you showed, manually? Where did you get them from?

**Reply by EdoFro** [reply]
No. they were all there.
Did you look into `Plugin-Einstellungen`?

**Reply by nnako** [reply]
Thanks again. I figured out that I was using v2.16.8 (from 2020). Now installed the current v2.16.22 and I see the plugins you mentioned :-) .

**Reply by nnako** [reply]
Hi @EdoFro, two questions I still have about the application of WinMerge's Prediff-Line-Filters on mindmap files. Maybe you have an explanation / solution?

**1. regex format**

you wrote your regexes in the following format:
```
\bFOLDED=".*"\B
```

from this I read

1. "any word separator" then
3. the literal string "FOLDED" then
4. a equal sign then
5. a double quote sign then
6. any further characters (.*) then
7. a double quote sign then
8. "NO word separator"

but I don't understand why you would specify (6) in the displayed way as it should include the next following double quote sign and why you would specify (8) as it shouldn't make any difference for the term, when replacing with empty content. would a regex like this be more clear?:
```
\bFOLDED="[a-z]*"
```

it regards the expected string format (in this case, booleans are represented only by "true" or "false" in lower-case letters) and I don't have to wrap my head around the closing "\B". Anyway, the result seems to be identical ;-) . Is there a particular reason you chose to specify your regex terms in that upper way?


**2. results in WinMerge**

![image](https://user-images.githubusercontent.com/5845602/187023057-b758d7ad-5e17-4a9e-b86d-26813e40803d.png)

in the image you see the following effect, twice, while my PrediffLineFilter is set up to filter the attributes `ID`, `FOLDED` and `MODIFIED` in the above mentioned way:

- in the left map, there is no `FOLDED` attribute, the lines for `ID` and `MODIFIED` are ignored - fine.
- in the right map, `ID` and `FOLDED` are ignored, but **not** `MODIFIED` in the first place - **?**
- in the right map, `ID` is shown as **not-ignored** even though the value is identical to the left side - **?**

WinMerge seems to match all ignored lines between the two sides without checking their content. this seems to lead to detected changes as soon as the number of ignored lines differs within the compared files of the current comparison scope.

Have you had this effect? Have you already thought about a solution for the whole block to be ignored as the `TEXT` content is identical between both compared maps? Currently WinMerge shows **two changes** within this context.

**Reply by EdoFro** [reply]
### 1. regex format

> Is there a particular reason you chose to specify your regex terms in that upper way?

yes!! : ignorance 😄 (I have no regex expertise)

### 2. results in WinMerge

I don't remember, but maybe I modified the parameters of the XML prettification step:

![image](https://user-images.githubusercontent.com/35700575/187087059-6c589127-f21c-4740-8d4e-7a24b7cca642.png)

the parameters in my computer are:
`-xml -indent --indent-attributes yes --tab-size 4 --indent-spaces 4 -wrap 0 --sort-attributes alpha`

**Reply by EdoFro** [reply]
I got a result similar to yours when unchecking "**Ignore blank lines**" in `Options / Compare / General`

![image](https://user-images.githubusercontent.com/35700575/187087615-08ab6ce6-533e-4be2-92e9-854aac33d813.png)

**Reply by EdoFro** [reply]
here is my exported Options .ini file so you can compare it with yours:

(exported using the "Export..." button from previous image)
[winmergeOptionsEdoFro.zip](https://github.com/freeplane/freeplane/files/9440180/winmergeOptionsEdoFro.zip)

**Reply by nnako** [reply]
the plugin arguments for PrettifyXML were unchanged. the `ignore blank lines` option did solve the issue. thanks, @EdoFro
